### PR TITLE
Requestify scoped import validation

### PR DESCRIPTION
**Pull request by hamishknight**
Move the lookup and validation of scoped imports into a request, and force the request when we're type-checking a primary file. This avoids having to do an additional pass during name binding, and has the nice bonus of no longer running the validation for secondary files.

This PR also adds `ModuleDecl::getTopLevelModule`, and uses it for the lookup of the scoped import rather than attempting to retrieve a module with the same name as the top-level module. This allows us to correctly retrieve the top-level module for a Clang submodule, rather than potentially retrieving the Swift module in a mixed source project.

Resolves SR-12265.

### CODE REVIEW

**Comment by hamishknight** [comment]
@swift-ci please test

File: test/ClangImporter/MixedSource/submodule.swift [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 @_exported import Mixed
 @_exported import Mixed.Submodule
+import func Mixed.Submodule.fromSubmodule
```
Comment (brentdax) [inline test/ClangImporter/MixedSource/submodule.swift]: Is this an addition to the language, or is it a bug fix? (Ask Doug if you aren’t sure.)

Does it preclude adding scoped nested imports later? There’s a lot of code in the compiler that seems oriented towards supporting that eventually.

**Review by beccadax — COMMENTED** [review]
Nice cleanup.

File: test/ClangImporter/MixedSource/submodule.swift (hunk 1 above)
Comment (hamishknight) [reply test/ClangImporter/MixedSource/submodule.swift]: @brentdax Thanks for the review! This is a bug fix (sorry, I forgot to add that to the PR description), we currently already support scoped imports on a submodule, e.g:

https://github.com/apple/swift/blob/853c0f545b61f5165c7c49fb24e675f305bdc760/test/IDE/Inputs/FooOverlay.swift#L3

We just didn't correctly handle the case where you're doing it for a submodule of the Clang part of a mixed source project.

I don't *think* this precludes adding scoped nested imports, though it'll probably be a bit tricky to implement if we continue to support scoped imports on submodules using this syntax. Ideally we'd require you to use some form of module qualification such as `Module.SubModule::Type`.

**Comment by hamishknight** [comment]
@swift-ci please test Linux platform

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

**Comment by hamishknight** [comment]
@swift-ci please test

**Comment by hamishknight** [comment]
@swift-ci please test source compatibility

File: test/NameBinding/scoped_import_overlay.swift [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// RUN: %empty-directory(%t)
+
+// RUN: %target-swift-frontend -enable-objc-interop -emit-module %S/Inputs/overlay.swift -module-name ClangModuleWithOverlay -I %S/Inputs/ClangModuleWithOverlay -o %t
+// RUN: %target-swift-frontend -typecheck %s -I %t
+
+// Make sure we can perform a scoped import on the overlay.
+import func ClangModuleWithOverlay.fromSwiftOverlay
+
+// ... as well as the underlying Clang module.
```
Comment (CodaFi) [inline test/NameBinding/scoped_import_overlay.swift]: Do we have any tests for doing this with import-as-member?

**Review by CodaFi — APPROVED** [review]
Great cleanup.  Only one question

File: test/NameBinding/scoped_import_overlay.swift (hunk 2 above)
Comment (hamishknight) [reply test/NameBinding/scoped_import_overlay.swift]: @CodaFi Not that I can find, though we don't currently support scoped imports for anything other than top-level decls. I'll add a couple of negative test cases for that.

Thanks for the review!

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Comment by hamishknight** [comment]
@compnerd Mind triggering again?

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please smoke test

**Comment by hamishknight** [comment]
@swift-ci please test Windows platform

**Comment by hamishknight** [comment]
@swift-ci please test Windows platform

**Comment by hamishknight** [comment]
@swift-ci please test Windows platform

**Comment by hamishknight** [comment]
@swift-ci please test Windows platform

**Comment by hamishknight** [comment]
@swift-ci please test Windows platform

**Pull request merged**